### PR TITLE
Add Trinity module

### PR DESCRIPTION
**Pull request by timslittle**
## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/trinity/main.nf [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/trinity:2.13.2--h00214ad_1':
+        'quay.io/biocontainers/trinity:2.13.2--h00214ad_1' }"
+
+    input:
+    tuple val(meta), path(reads)
+
+    output:
+    tuple val(meta), path("*Trinity.fasta"), emit: transcript_fasta
+    path "versions.yml"           , emit: versions
```
Comment (nvnieuwk) [inline modules/nf-core/trinity/main.nf:15]:
```suggestion
    tuple val(meta), path("*Trinity.fasta"), emit: transcript_fasta
    path "versions.yml"                    , emit: versions
```

File: modules/nf-core/trinity/main.nf [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    } else {
+        avail_mem = task.memory.giga
+    }
+
+    """
+    Trinity \\
+    --seqType ${seqType_args} \\
+    --max_memory ${avail_mem}G \\
+    ${reads_args} \\
+    --output ${prefix}_trinity \\
```
Comment (nvnieuwk) [inline modules/nf-core/trinity/main.nf:46]: It's preffered to not hard code your output file name, see here paragraph 6: https://nf-co.re/docs/contributing/modules#naming-conventions

File: modules/nf-core/trinity/main.nf [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    --CPU $task.cpus \\
+    $args
+
+    # Note that Trinity needs the word 'trinity' in the outdir
```
Comment (nvnieuwk) [inline modules/nf-core/trinity/main.nf:50]: Ah I see why you hardcoded it now, why does it need to have trinity in it? Seems a bit weird to me?

File: modules/nf-core/trinity/meta.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+input:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - reads:
+      type: file
+      description: fasta/fastq file
```
Comment (nvnieuwk) [inline modules/nf-core/trinity/meta.yml:25]: Can you extend this description a bit more?

File: modules/nf-core/trinity/meta.yml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  #
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - bam:
```
Comment (nvnieuwk) [inline modules/nf-core/trinity/meta.yml:39]:
```suggestion
  - transcript_fasta:
```

**Review by nvnieuwk — CHANGES_REQUESTED** [review]

File: modules/nf-core/trinity/main.nf (hunk 3 above)
Comment (timslittle) [reply modules/nf-core/trinity/main.nf:50]: I agree! But Trinity throws an error if you don't. Would it be better if we `mv`'d the output to a new directory?

Comment (nvnieuwk) [reply modules/nf-core/trinity/main.nf:50]: I think that would be the preffered way for most users :)

Comment (timslittle) [reply modules/nf-core/trinity/main.nf]: I think I got this working now!

Comment (nvnieuwk) [reply modules/nf-core/trinity/main.nf]: Allright looks good now! Can you fix the merge conflict first too?

**Review by nvnieuwk — APPROVED** [review]
LGTM! :rocket:

**Comment by nvnieuwk** [comment]
@nf-core-bot fix linting

**Comment by IdoBar** [comment]
The Singularity container currently used causes a segmentation fault error.
See https://github.com/nf-core/denovotranscript/issues/20

**Pull request merged**